Windows: fix echo for nmake

levitte opened the pull request:
It seems that nmake first tries to run executables on its own, and
only pass commands to cmd if that fails.  That means it's possible to
have nmake run something like 'echo.exe' when the builtin 'echo'
command was expected, which might give us unexpected results.

To get around this, we create our own echoing script and call it
explicitly from the nmake makefile.

Fixes #6670

levitte committed:
Windows: fix echo for nmake

It seems that nmake first tries to run executables on its own, and
only pass commands to cmd if that fails.  That means it's possible to
have nmake run something like 'echo.exe' when the builtin 'echo'
command was expected, which might give us unexpected results.

To get around this, we create our own echoing script and call it
explicitly from the nmake makefile.

Fixes #6670

levitte commented:
Do we have reasons to believe this is needed for 1.0.2 as well?

dot-asm commented on Configurations/windows-makefile.tmpl:
```diff
@@ -425,8 +427,8 @@ install_dev:
 install_engines:
-	@if "$(INSTALLTOP)"=="" ( echo INSTALLTOP should not be empty & exit 1 )
-	@echo *** Installing engines
+	@if "$(INSTALLTOP)"=="" ( $(ECHO) "INSTALLTOP should not be empty" & exit 1 )
```
On side note one doesn't have to replace *this* echo. Because it's beyond nmake's control. I mean nmake will have to pass this whole line to cmd.

dot-asm replied:
Ooooooh! It first looks for if.com, if.exe, if.bat, if.vbs, if.js, etc. before it decides to pass it to cmd. This actually feels like a bug...

dot-asm replied:
And if I try to (if ...), it look for (if.com, (if.exe, etc. It's just no escape...

dot-asm replied:
There is an escape! And if spells white space. I mean if there space right after @, then cmd is called. But it's not general rule. In sense that if I type `@ echo`, it still looks for echo.com, echo.exe, etc...

dot-asm replied:
> There is an escape!

Or not. It's actually not white space but a character that would be viewed by nmake as shell. Most notably & triggers call to cmd. So that one can shortcut `@rem ...` as `@rem & ...` One can also shortcut `@echo something` as `@echo something & :`

dot-asm commented:
To summarize.

As for `@rem {- output_on/off -}` one can either `@rem & {- output_on/off -}` or `@{- output_on/off}`.

As for `echo` one can replace with $(ECHO) or one can settle for `echo whatever & :`. I'd go for $(ECHO), because it gives better control and one doesn't have to clutter something that can be viewed as "simplest thing ever" with an idiom.

One should keep in mind that `if`s that don't have special characters will trigger search for if.com, if.exe, etc. One choose to terminate them with `& :`...

bernd-edlinger commented:
or if.pl ?

dot-asm commented:
> or if.pl?

Does it look for .pl? Well, it's surely possible to extent the list of possible executable extensions, so that it would be looking even for if.pl. It doesn't really makes nmake's behaviour less obnoxious, does it? :-) :-) :-)

bernd-edlinger commented:
Like util/if.pl maybe,
or preceding everything with %COMSPEC% /c ?
or would || or && be understood by cmd.exe, works mostly like if in bash?

dot-asm commented:
Editorial note. Commit message says "nmake first tries to **run** executables on its own, and
only pass commands to cmd if that fails." It's confusing because it suggest that it it actually executes something, sees if command failed, and re-tries with cmd. It can be fixed with "if it fails to find one on the %PATH%".

dot-asm commented:
> Like util/if.pl maybe,

That would pose strange problems. For example how would if "foo"=="" be passed to external program? As if foo==...

dot-asm commented:
> || or &&

Yes, that would work too...

richsalz commented:
So maybe, ignore nmake for now and focus on the things the 1.1.1 release needs?

levitte commented:
Err, @richsalz, `nmake` is the make implementation we support for MSVC builds.  If `nmake`'s behavior can potentially harm the builds / installs, it surely is something we do need to fix, no?

richsalz commented:
sorry, i got confused about which make we were using.  thanks.

dot-asm commented:
BTW, I'm actually find myself confused by & vs. &&. I mean I'm not actually sure what single ampersand mean. `rem & ...` surely works, but `& :` might be formally incorrect...

dot-asm commented:
> Do we have reasons to believe this is needed for 1.0.2 as well?

No. There is only one echo and it doesn't have wildchars. And there are no rems...

bernd-edlinger commented:
I thought & translates to ; execute sequentially
and && , || might work like in bash...

levitte commented:
Yes, that's correct, single ampersand is the oneline command separator with cmd.

bernd-edlinger commented:
maybe can we avoid `rem`, and live with the dangers of `if` ?
in the end it is a blatant nmake bug.

bernd-edlinger commented:
A # -style comment would not work in nmake?

levitte commented:
I just tested, it does look like `nmake` does the right thing with command lines that start with `#` (i.e. ignores them)

bernd-edlinger commented:
great!

richsalz commented:
how about change the stars to a non-wildcard character?

dot-asm commented:
> I just tested, it does look like nmake does the right thing with command lines that start with # (i.e. ignores them)

I can't confirm this.
```
all:
        # comment
        echo foo
```
says "'#' is not recognized as in internal or external command"

bernd-edlinger commented:
try this:

```
all:
# comment
          echo foo

```

bernd-edlinger commented:
> how about change the stars to a non-wildcard character?

works for me.

dot-asm commented:
> try this [# in beginning of line]

Seems to work. However, I'd find such formatting unfortunate, because it's not what you'd customarily expect. Well, one can always say "it's auto-generated, so why even look?" ...

dot-asm commented:
> how about change the stars to a non-wildcard character?

It's more than appropriate to get echo working in either case. Today it's wildcard, tomorrow something else...

bernd-edlinger commented:
Do you want to apply that to 1.1.0 as well?

dot-asm commented:
I'd like to see rem fixed at once. As already said, my suggestion as `rem & {- output_on/off -}` or as `{- output_on/off -}`. There also was suggestion for jagged `# {- output_on/off -}`, i.e. from beginning of line.

dot-asm commented:
A variant of `rem &` can be `rem {- output_on/off; "&" -}`...

bernd-edlinger commented:
I really don't see why everything needs to be fixed on one PR.

levitte committed:
Windows: avoid using 'rem' in the nmake makefile

To avoid the possibility that someone creates rem.exe, rem.bat or
rem.cmd, simply don't use it.  In the cases it was used, it was to
avoid empty lines, but it turns out that nmake handles those fine, so
no harm done.

levitte commented:
I saw no harm in adding that modification as well, so here goes.

dot-asm commented:
On closing note, let's try to remember that it occasionally looks for if.com, if.exe, etc. So that if anybody reports some incomprehensible error, one should ask for `where if` output.

levitte commented:
Agreed

levitte commented:
Merged.

master / 1.1.1:
1b6a0a261e22eb5a574bdb75da208817ffa2fbba Windows: avoid using 'rem' in the nmake makefile
9abce88b4b0055d6238a838aa00360152e185f02 Windows: fix echo for nmake

1.1.0:
5c06c7d748161697642bd4d1369acad4297abb3c Windows: avoid using 'rem' in the nmake makefile
566e333e52998e478ac6928c5d06bd89f4b37aba Windows: fix echo for nmake
